src/responder.ts

```diff
@@ -87,8 +87,16 @@ export abstract class EthereumResponder extends Responder {
     // implementations should query the provider (or a service) to figure out the appropriate gas price
     protected gasPrice = new ethers.utils.BigNumber(21000000000);
 
+    protected provider: ethers.providers.Provider;
+
     constructor(public readonly signer: ethers.Signer) {
         super();
+
+        if (!signer.provider) {
+            throw new ApplicationError("The given signer is not connected to a provider");
+        } else {
+            this.provider = signer.provider;
+        }
```

Comment:
I was always assuming that the provider is available on a Signer, but it's not the case; thus, now I'm checking in the constructor.

Reply:
Seems sensible. You could use ArgumentError here. Also, if you want to, you can make this a little more concise by not using the else here:

   if (!signer.provider) throw new ApplicationError("The given signer is not connected to a provider");
   this.provider = signer.provider;

Reply:
It also seems that EthereumResponder no longer requires the provider - I think this is since you refactored the submitStateFunction to be prepareRequest.

Reply:
Done.

EthereumResponder does not directly use the provider, but I expect any subclass to use it, so I'd leave it there as a protected member for now.